### PR TITLE
[YAML] - Kafka write and RAW format

### PR DESCRIPTION
**Pull request by ffernandez92**
addresses #28664

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by ffernandez92** [comment]
This PR completes the write Kafka part as well as the RAW feature. 

I ran a quick test using DF to validate that everything works as expected (in addition of course to the unit tests). It works fine.

Screenshot 2023-10-27 at 11 41 12

This is the YAML config I've used (writing here for documenting purposes):

```
pipeline:
  transforms:
    - type: ReadFromKafka
      config:
        topic: input_topic
        format: RAW
        bootstrap_servers: <kafka_bootstrap_servers>
    - type: WriteToKafka
      config:
        topic: output_topic
        format: RAW
        bootstrap_servers: <kafka_bootstrap_servers>
      input: ReadFromKafka.output

**Comment by ffernandez92** [comment]
Run Python_Coverage PreCommit

**Comment by ffernandez92** [comment]
R: @brucearctor

**Comment by robertwb** [comment]
Looks like all tests are passing now. 

As for documentation, yeah, I've been putting things in markdown files right in the directory as a start before we're ready to call this stable (which is coming up with 2.52). I've filed https://github.com/apache/beam/issues/29165 which could probably have sub-issues filed.

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java [hunk 1]
```diff
@@ -131,10 +134,18 @@ public void finish() {
     @Override
     public PCollectionRowTuple expand(PCollectionRowTuple input) {
       Schema inputSchema = input.get("input").getSchema();
-      final SerializableFunction<Row, byte[]> toBytesFn =
-          configuration.getFormat().equals("JSON")
-              ? JsonUtils.getRowToJsonBytesFunction(inputSchema)
-              : AvroUtils.getRowToAvroBytesFunction(inputSchema);
+      final SerializableFunction<Row, byte[]> toBytesFn;
+      if (configuration.getFormat().equals("RAW")) {
+        if (!inputSchema.hasField(PAYLOAD)) {
```
Comment (robertwb) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java:139]: In PubSub, I've been just requiring a single payload (not caring about the name). Do we want to have the same pattern here?

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java [hunk 2]
```diff
@@ -193,7 +217,7 @@ public abstract static class KafkaWriteSchemaTransformConfiguration implements S
     @SchemaFieldDescription(
         "A list of host/port pairs to use for establishing the initial connection to the"
             + " Kafka cluster. The client will make use of all servers irrespective of which servers are specified"
-            + " here for bootstrapping—this list only impacts the initial hosts used to discover the full set"
+            + " here for bootstrappingâ€”this list only impacts the initial hosts used to discover the full set"
```
Comment (robertwb) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java:220]: Revert?

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java (hunk 1 above)
Comment (ffernandez92) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java:139]: Yes, that's a tricky one. Kafka doesn't have the same attribute mechanism as PubSub. Maybe what we can do is allow the user to decide which "field" they want to pull from the Row. Then, we could have something like this:

```
    if len(field_names) != 1:
      raise ValueError(f'Expecting exactly one field, found {field_names}')
```
^ From the PubSub implementation. Then we do the input.getBytes(name_of_the_field_that_want_to_pull). I guess we need to come up with a good name for that like, getRowField. What do you think?

Comment (robertwb) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java:139]: If there's exactly one field we don't have to ask what it's name is, we can get it at pipeline construction time. (As another data point, this is what I did for write to text as well.)

Comment (ffernandez92) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java]: Yep, you're right. I pushed the necessary changes

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java [hunk 3]
```diff
@@ -131,10 +132,18 @@ public void finish() {
...
     public PCollectionRowTuple expand(PCollectionRowTuple input) {
       Schema inputSchema = input.get("input").getSchema();
-      final SerializableFunction<Row, byte[]> toBytesFn =
-          configuration.getFormat().equals("JSON")
-              ? JsonUtils.getRowToJsonBytesFunction(inputSchema)
-              : AvroUtils.getRowToAvroBytesFunction(inputSchema);
+      final SerializableFunction<Row, byte[]> toBytesFn;
+      if (configuration.getFormat().equals("RAW")) {
+        int numFields = inputSchema.getFields().size();
+        if (numFields != 1) {
```
Comment (robertwb) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java:138]: Perhaps check its type as well?

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProviderTest.java [hunk 4]
```diff
@@ -47,6 +50,9 @@ public class KafkaWriteSchemaTransformProviderTest {
 
   private static final Schema BEAMSCHEMA =
       Schema.of(Schema.Field.of("name", Schema.FieldType.STRING));
+
+  private static final Schema BEAMRAWSCHEMA =
```
Comment (robertwb) [inline sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProviderTest.java:54]: Nit: Why aren't these CAP_UNDERSCORE_CASE?

**Review by robertwb — APPROVED** [review]
Thanks!

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java (hunk 3 above)
Comment (robertwb) [reply sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaWriteSchemaTransformProvider.java:138]: We can do this as a follow-up.

**Pull request merged**